### PR TITLE
Multiobjective retrofit optimization analysis

### PR DESCRIPTION
**Pull request by nunezco2**
This code integrates the multiobjective retrofit optimization analysis by Yunjie and collaborators.

* Code was componentized into functions that implement the analysis and separate responsibilities
* Code only generates three output files insteads of >100, removing IO bounds from performance
* Code reproduces results of original notebook results

### CODE REVIEW

**Comment by nunezco2** [comment]
Here is the original notebook for review
[MultiObjectiveModel_Joplin_Pyomo.ipynb.zip](https://github.com/IN-CORE/pyincore/files/7672107/MultiObjectiveModel_Joplin_Pyomo.ipynb.zip)

**Comment by navarroc** [comment]
Can you please add this new analysis to pyincore/docs/source/modules.rst so the documentation is picked up?

**Comment by navarroc** [comment]
Or maybe @diegoac2 did you want to add the analysis to modules.rst? Should be a small change.

**Comment by navarroc** [comment]
I ran the test class and got an error that it hit an infeasible point, is that expected? To run this, I copied the code into my existing conda install so it's possible I messed something up.

`****OPTIMIZING BUILDING FUNCTIONALITY SUBJECT TO ECONOMIC LOSS AND POPULATION DISLOCATION EPSILON CONSTRAINTS****
Step  1  e:  2156.577171784068   e2: 34227.90332901029
WARNING - PyomoModel.py:load_from() - Loading a SolverResults object with a warning status into model.name="unknown";
  - termination condition: infeasible
  - message from solver: Ipopt 3.14.4\x3a Converged to a locally infeasible point. Problem may be infeasible.
infeasible
INFO - infeasible.py:log_infeasible_constraints() - CONSTR retrofit_budget_constraint: 192.36562836451054 </= 181.40803393552002
INFO - infeasible.py:log_infeasible_constraints() - CONSTR objective_1_constraint: 2156.577193349836 </= 2156.577171784068
INFO - infeasible.py:log_infeasible_constraints() - CONSTR objective_2_constraint: 34227.9034290102 </= 34227.90332901029
Empty DataFrame
Columns: [Economic Loss(Million Dollars), Dislocation Value, Functionality Value]
Index: []
Infeasible rows dropped:  0  rows.
Elapsed time:  259.2495958805084
Traceback (most recent call last):
  File "test_multiobjectiveretrofitoptimization.py", line 25, in <module>
    run_base_analysis()
  File "test_multiobjectiveretrofitoptimization.py", line 21, in run_base_analysis
    retrofit_optimization.run_analysis()
  File "/home/cnavarro/miniconda3/envs/pyincore-1.2.0-rc1/lib/python3.8/site-packages/pyincore/baseanalysis.py", line 264, in run_analysis
    return self.run()
  File "/home/cnavarro/miniconda3/envs/pyincore-1.2.0-rc1/lib/python3.8/site-packages/pyincore/analyses/multiobjectiveretrofitoptimization/multiobjectiveretrofitoptimization.py", line 82, in run
    self.multiobjective_retrofit_optimization_model(model_solver, num_epsilon_steps, budget_available,
  File "/home/cnavarro/miniconda3/envs/pyincore-1.2.0-rc1/lib/python3.8/site-packages/pyincore/analyses/multiobjectiveretrofitoptimization/multiobjectiveretrofitoptimization.py", line 126, in multiobjective_retrofit_optimization_model
    self.set_result_csv_data("out1", file_list[0], name="out1",
  File "/home/cnavarro/miniconda3/envs/pyincore-1.2.0-rc1/lib/python3.8/site-packages/pyincore/baseanalysis.py", line 231, in set_result_csv_data
    dataset = Dataset.from_dataframe(result_data, name, dataset_type)
  File "/home/cnavarro/miniconda3/envs/pyincore-1.2.0-rc1/lib/python3.8/site-packages/pyincore/dataset.py", line 123, in from_dataframe
    dataframe.to_csv(name, index=False)
AttributeError: 'str' object has no attribute 'to_csv'`

File: pyincore/analyses/multiobjectiveretrofitoptimization/multiobjectiveretrofitoptimization.py [hunk 1]
```diff
@@ -0,0 +1,1343 @@
...
+                    'description': 'A csv file with strategy cost data'
+                                   'per building',
+                    'type': 'incore:multiobjectiveStrategyCosts'
+                },
+            ],
+            'output_datasets': [
+                {
+                    'id': 'out1',
+                    'parent_type': '',
+                    'description': 'out1',
```
Comment (navarroc) [inline pyincore/analyses/multiobjectiveretrofitoptimization/multiobjectiveretrofitoptimization.py:1327]: Can we provide a better description of these outputs?

File: pyincore/analyses/multiobjectiveretrofitoptimization/multiobjectiveretrofitoptimization.py [hunk 2]
```diff
@@ -0,0 +1,1343 @@
...
+            | Science: Charles Nicholson, Yunjie Wen
+            | Implementation: Dale Cochran , Tarun Adluri , Jorge Duarte, Santiago Núñez-Corrales,
+                              and NCSA IN-CORE Dev Team
+
+        Related publications
+
+
+        Args:
+            incore_client (IncoreClient): Service authentication.
+        """
```
Comment (ncsa-mo) [inline pyincore/analyses/multiobjectiveretrofitoptimization/multiobjectiveretrofitoptimization.py:40]: A space after Args, before """


        Args:
            incore_client (IncoreClient): Service authentication.

        """

File: pyincore/analyses/multiobjectiveretrofitoptimization/multiobjectiveretrofitoptimization.py [hunk 3]
```diff
@@ -0,0 +1,1343 @@
...
+
+        Args:
+            model_solver (str): model solver to use for analysis
+            num_epsilon_steps (int): number of epsilon values for the multistep optimization algorithm
+            budget_available (float): budget constraint of the optimization analysis
+            scaling_factor (float): scaling factor for Q and Sc matrices
+            inactive_submodels (list): submodels to avoid during the computation
+            building_functionality_csv (pd.DataFrame): building repairs after a disaster event
+            strategy_costs_csv (pd.DataFrame): strategy cost data per building
+        """
```
Comment (ncsa-mo) [inline pyincore/analyses/multiobjectiveretrofitoptimization/multiobjectiveretrofitoptimization.py:99]: space, we use Google style, see https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: pyincore/analyses/multiobjectiveretrofitoptimization/multiobjectiveretrofitoptimization.py [hunk 4]
```diff
@@ -0,0 +1,1343 @@
...
+
+    def configure_model(self, budget_available, scaling_factor, building_functionality_csv, strategy_costs_csv):
+        """ Configure the base model to perform the multiobjective optimization.
+
+        Args:
+            budget_available (float): available budget
+            scaling_factor (float): value to scale monetary input data
+            building_functionality_csv (DataFrame): table containing building functionality data
+            strategy_costs_csv (DataFrame): table containing retrofit strategy costs data
+        Returns
```
Comment (ncsa-mo) [inline pyincore/analyses/multiobjectiveretrofitoptimization/multiobjectiveretrofitoptimization.py:142]: space before Returns (after Args:), and after Returns
```
            strategy_costs_csv (DataFrame): table containing retrofit strategy costs data

        Returns
            ConcreteModel: a base, parameterized cost/functionality model.

        """
```

**Comment by navarroc** [comment]
With the latest changes, this runs as expected. We just need a few documentation items addressed (once the review is finished by the researchers) and some minor spacing that Michal noticed and then we can merge this final PR.

**Comment by ncsa-mo** [comment]
This goes to the pyincore/docs/source/modules.rst

```
analyses/multiobjectiveretrofitoptimization
===========================================
..  autoclass:: multiobjectiveretrofitoptimization.multiobjectiveretrofitoptimization.MultiObjectiveRetrofitOptimization
    :members:

```

**Review by navarroc — APPROVED** [review]

**Review by ncsa-mo — APPROVED** [review]

**Review by ywkim312 — APPROVED** [review]

**Pull request merged**